Code/Tools/AssetProcessor/native/FileWatcher/FileWatcher.cpp

```diff
@@ -6,154 +6,119 @@
  *
  */
 #include "FileWatcher.h"
+#include "AzCore/std/containers/vector.h"
 #include <native/assetprocessor.h>
+#include <native/FileWatcher/FileWatcher_platform.h>
+#include <QFileInfo>
 
-//////////////////////////////////////////////////////////////////////////////
-/// FolderWatchRoot
-void FolderRootWatch::ProcessNewFileEvent(const QString& file)
+//! IsSubfolder(folderA, folderB)
+//! returns whether folderA is a subfolder of folderB
+//! assumptions: absolute paths, case insensitive
+static bool IsSubfolder(const QString& folderA, const QString& folderB)
 {
-    FileChangeInfo info;
-    info.m_action = FileAction::FileAction_Added;
-    info.m_filePath = file;
-    const bool invoked = QMetaObject::invokeMethod(m_fileWatcher, "AnyFileChange", Qt::QueuedConnection, Q_ARG(FileChangeInfo, info));
-    Q_ASSERT(invoked);
-}
+    // lets avoid allocating or messing with memory - this is a MAJOR hotspot as it is called for any file change even in the cache!
```

Comment:
Eventually, it may be worth using `std::filesystem::paths` instead of QStrings everywhere. This lets us have a platform-native representation of the path. Then for routines like this, we could just `std::search` to iterate each one and compare components. e.g. https://stackoverflow.com/questions/62503197/check-if-path-contains-another-in-c

Reply:
@jeremyong-az We actually have our own [Path](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/Path/Path.h#L18-L40) class that tries to model python pathlib, while also having path segment iteration support.

We also have an [AZ::IO::PathView::IsRelativeTo](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/Path/Path.h#L181-L182) function to determine if a path is contained within another

Reply:
I didn't use the `IsRelativeTo` function due to that previously-existing comment (this function was moved from another place):
```cpp
// lets avoid allocating or messing with memory - this is a MAJOR hotspot as it is called for any file change even in the cache!
```

I haven't done any profiling on this function myself to know how sensitive it is. But using `PathView::IsRelativeTo` would require copying both parameters.

Reply:
This is fixed in the latest commits.